src/envoy/http/path_rewrite/filter.cc

```diff
@@ -79,19 +79,11 @@ FilterHeadersStatus Filter::decodeHeaders(RequestHeaderMap& headers, bool) {
 
   // Make sure route is calculated
   auto route = decoder_callbacks_->route();
-  if (route == nullptr || route->routeEntry() == nullptr) {
-    config_->stats().denied_by_no_route_.inc();
 
-    rejectRequest(
-        Envoy::Http::Code::NotFound,
-        absl::StrCat("Request `", utils::readHeaderEntry(headers.Method()), " ",
-                     utils::readHeaderEntry(headers.Path()),
-                     "` is not defined by this API."),
-        utils::generateRcDetails(utils::kRcDetailFilterPathRewrite,
-                                 utils::kRcDetailErrorTypeUndefinedRequest));
-    decoder_callbacks_->streamInfo().setResponseFlag(
-        Envoy::StreamInfo::ResponseFlag::NoRouteFound);
-    return FilterHeadersStatus::StopIteration;
+  // If no route is matched, pass through the request and it will be rejected by
+  // router.
+  if (route == nullptr || route->routeEntry() == nullptr) {
```

Comment:
There are two options to return both 404 and 405.
1) handled in service_control:  if could not find a route,  check all_configured_paths,  found 405, not found 404
2) handled by router.  we need to create one route_entry for each path, to return 405,  and catch_all to return 404.

I prefer 1) for:
* we are in control of how errors are returned, 
* less number of route entries. faster for 404

Reply:
I still prefer to put in the router since it is more reasonable to add there in terms of filter functionality, instead of plug in our custom filters.

- *we are in control of how errors are returned*
We can still add static direct response for different NotFound requests
- *we can create one stats for each case*
No idea for now if we reject in router but on the other hand, we have a better trace annotation. 
- *less number of route entries. faster for 404*
Not real. Only the 404 and 405 requests will fall into these route entries so it won't effect the normal request handling.

Reply:
In the route direct_response error message, can we print request info, such as path?  e.g. can we do
"405: path: /foo/bar is supported in the API, but method: POST is not supported"

Reply:
Given that we're building the direct response, we can just put the `uri_template` as the error message. Not sure if we can put the request path, but URI template should be good enough.

"Path `/book/{shelf_id}` is supported, but method `POST` is not supported"

Reply:
In the direct_response route,  we could not print rejected request path and method.  The method could be any of POST, DELETE, PATCH, OPTIONS  if supported one is GET.
My point is:  if we handle the rejection in service_control, we have more flexible. 
This is inline with Josh request of moving jwt failure handling into service_control.

Reply:
I don't think we can pull the request info in the runtime. 

For the catchall route match configuration, We can only put the `path` in it considering the defined requests have been matched before, because
- it is impossible illustrate all the http methods including the custom one
- it can help us aggregate all the methods into several ones by paths

It can be `For path /book/{shelf_id}, the http method in the current request is not supported`.

Reply:
`This is inline with Josh request of moving jwt failure handling into service_control.`

IMHO, it is only for the request rejected before the service control filter, which doesn't mean we need to move all the rejections in the service control.

Reply:
OK, I pefer 1) approach,   but not strongly.   If you two prefer 2),  I am OK.  please go for it.

Hi  @TAOXUY , is this PR for both 404 and 405, or  just 404?

Reply:
This is the first part and only for 404. The followup would be 405.